diff --git a/node_modules/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/AnyTypeProvider.kt b/node_modules/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/AnyTypeProvider.kt
new file mode 100644
index 0000000..d8b952f
--- /dev/null
+++ b/node_modules/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/AnyTypeProvider.kt
@@ -0,0 +1,10 @@
+package expo.modules.kotlin.types
+
+object AnyTypeProvider {
+  @PublishedApi
+  internal val typesMap get() = AnyTypeCache.typesMap
+}
+
+inline fun <reified T> AnyTypeProvider.cachedAnyType(): AnyType? {
+  return AnyTypeCache.cachedAnyType<T>()
+}
diff --git a/node_modules/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/LazyKType.kt b/node_modules/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/LazyKType.kt
new file mode 100644
index 0000000..9a4c5e9
--- /dev/null
+++ b/node_modules/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/LazyKType.kt
@@ -0,0 +1,12 @@
+package expo.modules.kotlin.types
+
+import kotlin.reflect.KClass
+import kotlin.reflect.KType
+
+class LazyKType(
+  val kClass: KClass<*>,
+  val isNullable: Boolean = false,
+  private val kTypeProvider: (() -> KType)? = null
+) {
+  val kType: KType? get() = kTypeProvider?.invoke()
+}
diff --git a/node_modules/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/ReturnType.kt b/node_modules/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/ReturnType.kt
index edf2186..1258722 100644
--- a/node_modules/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/ReturnType.kt
+++ b/node_modules/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/ReturnType.kt
@@ -93,6 +93,10 @@ inline fun <reified T> toReturnType(): ReturnType {
 class ReturnType(
   private val converter: JSTypeConverter<*>
 ) {
+  constructor(kClass: kotlin.reflect.KClass<*>) : this(
+    getDirectConverter(kClass.java) ?: getIndirectConverter(kClass.java, null)
+  )
+
   fun convertToJS(value: Any?): Any? {
     return converter.convertToJS(value)
   }
